FileArchiveModeFactory - Extracted from FileTarget

snakefoot opened the pull request:
Attempt to resolve the issues #1670 and #1681, which is mostly caused not understanding how to configure the NLog Archive Logic (And also major refactoring of the archive logic in FileTarget).

I guess more effort could be applied to FileArchiveModeDynamicSequence.GenerateFileNameMask() to better recognize any date-format-output.

Enabled the unit-tests from #1788, but had to modify them, so they would work with Mono on Linux (Changed hardcoded windows-backslash to Path.Combine). The tests also suffered from confusing debug- and trace-level (Logged to debug-level and didn't expect it to reach trace-level).

**!! Breaking change !!**
If using ArchiveNumberingMode.Sequence (default) and not having configured an ArchiveFileName (default) and have enabled archiving (MaxArchiveFiles > 0). Then it will now enable dynamic-sequence mode, instead of strict-sequence-mode.

Dynamic-sequence-mode will use the same archive-filename-format (as strict-sequence-mode), but its filemask-wildcard for doing archive-cleanup is a little more greedy as it replaces the largest range of digits (must be more than one) with a wildcard (and not just the sequence-no).

This improves the out-of-box archive-cleanup, when the FileTarget.FileName-Layout includes date-renderer.

**New behavior for dynamice sequence-mode**
New behavior when ArchiveFileName does not contain sequence-no-pattern `(#}`:

- More adaptive filemask-wildcard used when doing archive-cleanup. Before it was just all files with the same file-extension. Now also using the actual LogEvent (Instead of Null-LogEvent), so the filemask for cleanup is a lot better (Doesn't need an isolated archive directory).
- Now delays archive cleanup until first write-operation occurs, instead of FileTarget.Initialize(). This allows archive-cleanup to work properly as it will use the actual LogEvent for rendering archive filemask for cleanup (Instead of Null-LogEvent).
  - This also removes extra overhead at startup where all FileTarget would scan their archive-folders.
  - This generates extra overhead at first new file write, where archive-cleanup occurs. But the archive-cleanup is skipped if new file write is caused by just having executed a file-archive-operation.

This improves the out-of-box archive-cleanup, when the ArchiveFileName-Layout includes date-renderer.

**Other changes:**
- Now delays the monitoring of the archive-folder until the first-write. And the filemask monitored is generated from the actual logevent instead of Null-LogEvent (More correct).
- ArchiveNumberingMode.Date now skips an unnecessary archive-cleanup when moving to new file after just having performed archive of the previous file (with cleanup).
- Attempt to handle that logging-folder and archive-folder is the same folder. Crude Path.GetDirectoryName-comparison that probably can be cheated.
- Attempt to handle file-creation-date-issue on Linux for archiving. This should improve how NLog performs archive-cleanup of dated-archives.
- Attempt to handle non-standard-path-usage, that is different from those returned by the standard filesystem.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nlog/nlog/1993)
<!-- Reviewable:end -->

304NotModified commented:
It's the limited silverlight api. 

> Targets\FileArchiveStyleFactory.cs(583,81): error CS1061: 'System.IO.FileInfo' does not contain a definition for 'CreationTimeUtc' and no extension method 'CreationTimeUtc' accepting a first argument of type 'System.IO.FileInfo' could be found (are you missing a using directive or an assembly reference?) 

There's maybe a helper for this already.

snakefoot commented:
@304NotModified Thanks that fixed the appveyor-build. The travis-build required some changes to the unit-tests. But now things are looking green. And I have implemented a fix for my worries about performing too many archive cleanups.

304NotModified commented on src/NLog/Internal/FileInfoExt.cs:
```diff
@@ -43,19 +43,16 @@ public static DateTime GetLastWriteTimeUtc(this FileInfo fileInfo)
             return fileInfo.LastWriteTimeUtc;
 #else
-            return fileInfo.LastWriteTime;
+            return fileInfo.LastWriteTime.ToUniversalTime();
```
isn't this an expensive method? (dunno how many it will be called)

snakefoot replied:
Well it only happens for silverlight-applications, and since the method should return Utc, then it would make sense to do it correct. But I will remove it, since it is not important to me.

304NotModified replied:
ow yes, missed the `!`.

I think this is then OK, iif the tests could work without UTC stuff

304NotModified commented on src/NLog/Targets/FileArchiveStyleFactory.cs:
```diff
@@ -0,0 +1,758 @@
+        /// File name which is used as template for matching and replacements. 
+        /// It is expected to contain a pattern to match.
+        /// </summary>
+        public string Template { get { return this.template; } }
```
could we move to auto properties? (less code is more)

snakefoot replied:
I just moved the class directly from FileTarget, but I can clean it up.

304NotModified replied:
Would be great if you would do that as I cannot see what's new or what's moved (easily). Thx in advance!

304NotModified commented on src/NLog/Targets/FileArchiveStyleFactory.cs:
```diff
@@ -0,0 +1,758 @@
+            }
+            else
+            {
+                _lastArchiveFilePath = string.Empty;
```
please move before the `if` (as both branches have ` _lastArchiveFilePath = string.Empty;` - or is this a  potential bug?)

snakefoot replied:
Removed the if-statement.

304NotModified commented on src/NLog/Targets/FileArchiveStyleFactory.cs:
```diff
@@ -0,0 +1,758 @@
+            }
+
+            if (existingArchiveFiles.Count > 1)
+                existingArchiveFiles.Sort((x, y) => (x.Date != y.Date && !x.HasSameFormattedDate(y.Date)) ? x.Date.CompareTo(y.Date) : x.Sequence.CompareTo(y.Sequence) != 0 ? x.Sequence.CompareTo(y.Sequence) : string.CompareOrdinal(x.FileName, y.FileName));
```
any idea if we could get this a bit more readable (a helper maybe?)

snakefoot replied:
Now calling static-method

304NotModified commented on src/NLog/Targets/FileArchiveStyleFactory.cs:
```diff
@@ -0,0 +1,758 @@
+            if (existingArchiveFiles.Count == 0 || existingArchiveFiles.Count < maxArchiveFiles)
+                yield break;
+
+            for (int i = 0; i < existingArchiveFiles.Count - maxArchiveFiles; ++i)
```
`i++` instead of `++i` ?

I'm curious, what's the benefit of `++i` in a for loop?

edit: it was already there I think

snakefoot replied:
Changed to i++

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -1990,21 +1539,12 @@ private string GetArchiveFileNamePattern(string fileName, LogEventInfo eventInfo
         /// <param name="ev">Log event that the <see cref="FileTarget"/> instance is currently processing.</param>
         /// <param name="upcomingWriteSize">The size in bytes of the next chunk of data to be written in the file.</param>
-        private void TryArchiveFile(string fileName, LogEventInfo ev, int upcomingWriteSize)
+        private bool TryArchiveFile(string fileName, LogEventInfo ev, int upcomingWriteSize)
```
please document the return value.

snakefoot replied:
Updated XML-docs

304NotModified commented on tests/NLog.UnitTests/Targets/FileTargetTests.cs:
```diff
@@ -2770,8 +2837,8 @@ public void MaxArchiveFilesWithDate_in_sameDir(bool changeCreationAndWriteTime)
                     {
-                        File.SetCreationTime(filePath, time);
-                        File.SetLastWriteTime(filePath, time);
+                        File.SetCreationTimeUtc(filePath, time);
```
is this really needed? NLog should also handle local times

snakefoot replied:
Leftovers from when I was debugging the failing unit-tests on Mono/Travis. Will undo the changes.

304NotModified replied:
👍

snakefoot commented:
> Please move all the classes to their own file :)

Have now moved the classes into their own individual file.

snakefoot commented:
@304NotModified Discovered that the "bugs" reported for the NLog-Archive-Logic is actually caused by the problems with understanding how to use the enum ArchiveNumberingMode together with ArchiveFileName-layout (Failing to use the **{#}** properly).

I have now modified this PR into improving the out-of-box experience, when using the NLog-Archive-Logic. And removed the bad behavior of doing cleanup of all files with same file-extension, when failing to use the **{#}** properly.

304NotModified commented:
Sounds great! Will include into 4.5 due to the large redactor. Is it OK if I merge #2000  first?

snakefoot commented:
> Is it OK if I merge #2000 first?

Sure no problem with that.

304NotModified commented:
@snakefoot 

> If using ArchiveNumberingMode.Sequence (default) and not having configured an ArchiveFileName (default) and have enabled archiving (MaxArchiveFiles > 0). Then it will now enable dynamic-sequence mode, instead of strict-sequence-mode.

> Dynamic-sequence-mode will use the same archive-filename-format (as strict-sequence-mode), but its filemask-wildcard for doing archive-cleanup is a little more greedy as it replaces the largest range of digits (must be more than one) with a wildcard (and not just the sequence-no).

> This improves the out-of-box archive-cleanup, when the FileTarget.FileName-Layout includes date-renderer.

Do you have an example for this? (e.g. what is changed). Need that for the news post :)

snakefoot commented:
> Do you have an example for this?

I guess the most simple example is that you can now get correct archive cleanup when just using the two FileTarget parameters `FileName` and `MaxArchiveFiles`, eventhough `FileName` depends on layout-renders (Like date, logger-name, log-level, etc.). Before it would try and cleanup according to rendering of Null-LogEvent.

The more advanced example is when using `archiveFileName` without `{0}`, then one will also get correct archive cleanup. Before it would try and cleanup anything matching the log-file-extension.

304NotModified commented:
fixes #2050

304NotModified commented:
@snakefoot any idea if this also fixes #1748?

snakefoot commented:
@304NotModified No don't think so, just scanned the code to see if there was any obvious bugs, but didn't find any.

snakefoot commented:
> Any idea if this also fixes #1748?

@304NotModified Just tested with the provided nlog.config. It rolls every day (as configured), so the result becomes:

> 01_AppName.0.log (Two days ago)
> 02_AppName.0.log (Previous day)
> 03_AppName.log (Active)

I think the wanted behavior is this:

> 01_AppName.log (Two days ago)
> 02_AppName.log (Previous day)
> 03_AppName.log (Active)

This PR can give this result by removing **ArchiveEvery="Day"** and **ArchiveNumbering="Rolling"**. And then change the existing parameter to **MaxArchiveFiles="27"** (Cater for february).

snakefoot commented:
@304NotModified Have now implemented a unit test that verifies that the above solution actually works for  #1748

304NotModified commented:
fixes  #1670, fixes #1681, fixes https://github.com/NLog/NLog/issues/1748
